File: FateGrandAutomata/HighlightView.cs
using System.Collections.Generic;
using Android.Content;
using Android.Graphics;
using Android.Views;
using Region = CoreAutomata.Region;

namespace FateGrandAutomata
{
    public class HighlightView : View
    {
        static readonly HashSet<Region> Regions = new HashSet<Region>();

        static HighlightView _instance;

        public static void AddRegion(Region Region)
        {
            _instance?.Post(() =>
            {
                Regions.Add(Region);

                _instance.Invalidate();
            });
        }

        public static void RemoveRegion(Region Region)
        {
            _instance?.Post(() =>
            {
                Regions.Remove(Region);

                _instance.Invalidate();
            });
        }

        readonly Paint _paint;

        public HighlightView(Context Context) : base(Context)
        {
            _instance = this;

            _paint = new Paint
            {
                Color = Color.Red,
                StrokeWidth = 5
            };

            _paint.SetStyle(Paint.Style.Stroke);
        }

        protected override void OnDraw(Canvas Canvas)
        {
            base.OnDraw(Canvas);

            foreach (var region in Regions)
            {
                Canvas.DrawRect(region.X, region.Y, region.R, region.B, _paint);
            }
        }
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/ui/HighlightView.kt
package com.mathewsachin.fategrandautomata.ui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.view.View
import com.mathewsachin.libautomata.Region

class HighlightView(
    Context: Context,
    val regionsToHighlight: Map<Region, Boolean>
) : View(Context) {
    private val red = Paint().apply {
        color = Color.RED
        strokeWidth = 5f
        style = Paint.Style.STROKE
    }

    private val green = Paint().apply {
        color = Color.GREEN
        strokeWidth = 5f
        style = Paint.Style.STROKE
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

        for ((region, success) in regionsToHighlight) {
            canvas?.drawRect(
                region.X.toFloat(),
                region.Y.toFloat(),
                region.right.toFloat(),
                region.bottom.toFloat(),
                if (success) green else red
            )
        }
    }
}